app/src/main/java/net/aiscope/gdd_app/repository/SampleDtos.kt

```diff
@@ -22,25 +24,42 @@ data class SampleDto(
     @SerializedName("microscopeQuality") val microscopeQuality: MicroscopeQualityDto?,
     @SerializedName("imagePaths") val imagePaths: List<String>,
     @SerializedName("maskPaths") val maskPaths: List<String>,
+    @SerializedName("areMasksEmpty") val areMasksEmpty: List<Boolean>,
     @SerializedName("metadata") val metadata: SampleMetadataDto,
     @SerializedName("status") val status: Short,
     @SerializedName("createdOn") val createdOn: Calendar = Calendar.getInstance(),
-    @SerializedName("lastModified") val lastModified : Calendar = Calendar.getInstance()
+    @SerializedName("lastModified") val lastModified: Calendar = Calendar.getInstance()
 ) {
-    fun toDomain(): Sample = Sample(
-        id = id,
-        healthFacility = healthFacility,
-        microscopist = microscopist,
-        disease = disease,
-        preparation = preparation?.toDomain(),
-        microscopeQuality = microscopeQuality?.toDomain(),
-        images = imagePaths.map { File(it) }.toLinkedHashSet(),
-        masks = maskPaths.map { File(it) }.toLinkedHashSet(),
-        metadata = metadata.toDomain(),
-        status = SampleStatus.values().first { it.id == status },
-        createdOn = createdOn,
-        lastModified = lastModified
-    )
+    fun toDomain(): Sample {
+        val areMasksEmpty = backfillAreMaskEmpty()
+        val completedCaptures = buildCompletedCaptures(areMasksEmpty)
+        val inProgressCapture = extractInProgressCapture()
+
+        return Sample(
+            id = id,
+            healthFacility = healthFacility,
+            microscopist = microscopist,
+            disease = disease,
+            preparation = preparation?.toDomain(),
+            microscopeQuality = microscopeQuality?.toDomain(),
+            images = Images(inProgressCapture, completedCaptures),
+            metadata = metadata.toDomain(),
+            status = SampleStatus.values().first { it.id == status },
+            createdOn = createdOn,
+            lastModified = lastModified
+        )
+    }
+
+    private fun backfillAreMaskEmpty() =
+        if (areMasksEmpty.isNotEmpty()) areMasksEmpty else List(imagePaths.size) { false }
```

Comment:
So this backfill is only for samples which have been made on a previous version of the app? Or is this load bearing?

Reply:
Yes, it's for the former. The edge-case that this tries to avoid is when the app is updated to this new version when there are still samples to upload. Without this backfill, due to the `zip` below, we would lose the images for upload when recovering the data from the repository.

Reply:
Cool 👍